sycl/include/sycl/esimd/esimd.hpp

```diff
@@ -0,0 +1,398 @@
+//==------------ - esimd.hpp - DPC++ Explicit SIMD API   -------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// Implement Explicit SIMD vector APIs.
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/esimd/detail/esimd_intrin.hpp>
+#include <sycl/esimd/detail/esimd_types.hpp>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace intel {
+namespace gpu {
+
+//
+// The simd vector class.
+//
+// This is a wrapper class for llvm vector values. Additionally this class
+// supports region operations that map to Intel GPU regions. The type of
+// a region select or format operation is of simd_view type, which models
+// read-update-write semantics.
+//
+template <typename Ty, int N> class simd {
+public:
+  // The underlying builtin data type.
+  using vector_type = vector_type_t<Ty, N>;
+
+  // The element type of this simd object.
+  using element_type = Ty;
+
+  // The number of elements in this simd object.
+  static constexpr int length = N;
+
+  // Constructors.
+  constexpr simd() = default;
+  constexpr simd(const simd &other) { set(other.data()); }
+  constexpr simd(simd &&other) { set(other.data()); }
+  constexpr simd(const vector_type &Val) { set(Val); }
```

Comment:
Are these constructors `noexcept` as well?

Reply:
Only constructors with full initialization pattern are currently specified with noexcept, as they have a straightforward implementation without throwing and are expected to be fully optimized by compiler. The other constructors are implemented differently based on host or device code, which may be subject to change.

Reply:
Ok, that makes sense.  Thanks for the explanation.